core/chains/terra/terratxm/txm.go

```diff
@@ -0,0 +1,355 @@
+	}
+	for txHash, msgs := range msgsByTxHash {
+		err := txm.confirmTx(txHash, getIDs(msgs))
+		if err != nil {
+			txm.lggr.Errorw("unable to confirm broadcasted but unconfirmed txes", "err", err, "txhash", txHash)
+		}
+	}
+}
+
+func (txm *Txm) run(sub pg.Subscription) {
+	defer close(txm.done)
+	txm.confirmAnyUnconfirmed()
+	for {
+		select {
+		case <-sub.Events():
+			txm.sendMsgBatch()
+		case <-txm.ticker.C:
+			txm.sendMsgBatch()
+		case <-txm.stop:
+			return
+		}
+	}
+}
+
+func (txm *Txm) sendMsgBatch() {
+	unstarted, err := txm.orm.SelectMsgsWithState(Unstarted)
+	if err != nil {
+		txm.lggr.Errorw("unable to read unstarted msgs", "err", err)
+		return
+	}
+	if len(unstarted) == 0 {
+		return
+	}
+	if len(unstarted) > MaxMsgsPerBatch {
+		unstarted = unstarted[:MaxMsgsPerBatch+1]
+	}
+	txm.lggr.Debugw("building a batch", "batch", unstarted)
+	var msgsByFrom = make(map[string][]TerraMsg)
+	for _, m := range unstarted {
+		var ms wasmtypes.MsgExecuteContract
+		err := ms.Unmarshal(m.Msg)
+		if err != nil {
+			// Should be impossible given the check in Enqueue
+			txm.lggr.Errorw("failed to unmarshal msg, skipping", "err", err, "msg", m)
+			continue
+		}
+		m.ExecuteContract = &ms
+		_, err = sdk.AccAddressFromBech32(ms.Sender)
+		if err != nil {
+			// Should never happen, we parse sender on Enqueue
+			txm.lggr.Errorw("unable to parse sender", "err", err, "sender", ms.Sender)
+			continue
+		}
+		msgsByFrom[ms.Sender] = append(msgsByFrom[ms.Sender], m)
+	}
+
+	txm.lggr.Debugw("msgsByFrom", "msgsByFrom", msgsByFrom)
+	gp := txm.tc.GasPrice(txm.fallbackGasPrice)
+	for s, msgs := range msgsByFrom {
+		sender, _ := sdk.AccAddressFromBech32(s) // Already checked validity above
+		key, err := txm.ks.Get(sender.String())
+		if err != nil {
+			txm.lggr.Errorw("unable to find key for from address", "err", err, "from", sender.String())
+			// We check the transmitter key exists when the job is added. So it would have to be deleted
+			// after it was added for this to happen. Retry on next poll should the key be re-added.
+			return
+		}
+		txm.sendMsgBatchFromAddress(gp, sender, key, msgs)
+	}
+}
+
+func (txm *Txm) sendMsgBatchFromAddress(gasPrice sdk.DecCoin, sender sdk.AccAddress, key terrakey.Key, msgs []TerraMsg) {
+	an, sn, err := txm.tc.Account(sender)
+	if err != nil {
+		txm.lggr.Errorw("unable to read account", "err", err, "from", sender.String())
+		// If we can't read the account, assume transient api issues and leave msgs unstarted
+		// to retry on next poll.
+		return
+	}
+
+	txm.lggr.Debugw("simulating batch", "from", sender, "msgs", msgs)
+	simResults, err := txm.tc.BatchSimulateUnsigned(getSimMsgs(msgs), sn)
+	if err != nil {
+		txm.lggr.Errorw("unable to simulate", "err", err, "from", sender.String())
+		// If we can't simulate assume transient api issue and retry on next poll.
+		return
+	}
+	txm.lggr.Debugw("simulation results", "from", sender, "succeeded", simResults.Succeeded, "failed", simResults.Failed)
```

Comment:
Is it always safe to simulate?

Reply:
for now yes, if we were to support other job types on terra we'd have to revisit this